### PR TITLE
[TwigComponent] - Dynamic template

### PR DESCRIPTION
**Pull request by seb-jean**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Tickets       | Fix https://github.com/symfony/ux-twig-component/pull/1
| License       | MIT

Hi,

I add this PR which has already been added here (https://github.com/symfony/ux-twig-component/pull/1) and I modified it a bit.

The idea is to have a dynamic template:

```php
<?php

namespace App\Components;

use Symfony\UX\TwigComponent\Attribute\AsTwigComponent;

#[AsTwigComponent('alert')]
class AlertComponent
{
    public string $message;
    public string $type = 'success';
    
    public function getTemplate() : ?string
    {
        return 'components/alert-'.$this->type.'.html.twig';
    }
}

```

### CODE REVIEW

**Comment by weaverryan** [comment]
Hey @seb-jean!

Thanks for bringing this over here! Does this work yet? I would expect that this would be configured on the attribute itself - e.g. `#[AsTwigComponent('alert'), template='templates/foo.html.twig']`

Cheers!

**Comment by seb-jean** [comment]
@weaverryan Yes PR works well.

**Comment by seb-jean** [comment]
In fact, I don't see how to get:
`#[AsTwigComponent('alert'), template='components/alert-{type}.html.twig']`

File: src/TwigComponent/src/ComponentRenderer.php [hunk 1]
```diff
@@ -30,6 +30,7 @@ public function __construct(Environment $twig)
     public function render(object $component, string $template): string
     {
         // TODO: Self-Rendering components?
-        return $this->twig->render($template, ['this' => $component]);
+        $templateName = $component->getTemplate() ?? $template;
```
Comment (weaverryan) [inline src/TwigComponent/src/ComponentRenderer.php:33]: I don't understand how this works (I could be totally missing something). The `$component` here is your component instance. But your component does not use or extend `AsTwigComponent`: that's an attribute on your component.

Comment (seb-jean) [reply src/TwigComponent/src/ComponentRenderer.php:33]: I didn't need to put `function getTemplate()` in `AsTwigComponent`. It works without.

**Comment by seb-jean** [comment]
Yes that might solve the problem :).

**Comment by kbond** [comment]
@seb-jean, are you up to modify this PR to a self-rendering implementation?

**Comment by seb-jean** [comment]
@kbond , no, I'm sorry, I don't see what to do next.

**Comment by kbond** [comment]
No worries, I'll create a PR when I have a chance.

**Comment by kbond** [comment]
@seb-jean what about https://github.com/symfony/ux/pull/166#issuecomment-970893084 as a solution?

**Comment by seb-jean** [comment]
@kbond, the solution works very well.
Thank you so much :pray:.

**Pull request closed without merging**